*,bazel: generate few more files within the sandbox

alan-mas opened the pull request:
Touches #57787, we convert over a few more stringer targets and other
instances of files generated using `//go:generate go run <...>`.
files within Bazel.

Of note in this (beast of a) commit is the generation of
roachpb/batch_generated.go, which much like langgen, makes use of
symbols in the package to generate the file later needed to compile the
package. To that end we introduce a stand-alone 'bootstrap' target.

For other files generated using `go run fname.go`, we convert them over
into standalone binaries.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59540)
<!-- Reviewable:end -->

irfansharif reviewed (commented):
Discussed internally but copying over for posterity: If I’m reading correctly, we're trying package out the work done in `pkg/util/interval/generic/gen.sh`. My instinct here is to use the simpler `sh_binary` instead of fully fleshing out a bazel-ified version of `go_generics`. We’re doing something similar in `pkg/sql/parser/BUILD.bazel` with `sql-gen.sh`.

irfansharif commented on pkg/util/timeutil/BUILD.bazel:
```diff
@@ -48,3 +49,48 @@ go_test(
+    outs = ["lowercase_timezones.go"],
+    cmd = """
+      env PATH=`dirname $(location @go_sdk//:bin/go)` HOME=$(GENDIR) \
+      go run $(location generate_lowercase_timezones.go) 
```
Is it not possible to simply do "$(location @go_sdk//:bin/go) run ..."?

alan-mas replied:
It is running either ways (with "$(location @go_sdk//:bin/go) run ..." and "go run") but it is failing due to:
`failed to run crlfmt: exec: "crlfmt": executable file not found in $PATH`

I can see `crlfmt` is kind of an internal library, right?

irfansharif replied:
It's a binary, https://github.com/cockroachdb/crlfmt. We'd have to specify it as a tool/binary dependency, in the same we declare `execgen` as a binary dependency before being allowed to use it.

alan-mas replied:
I already added inside `tools deps` (I think so):
```
    tools = [
        "@go_sdk//:bin/go",
        "@com_github_cockroachdb_crlfmt//:crlfmt",
    ],
    ```
    But I am not sure how I can add it in `$PATH` as that `crlfmt` is needed inside `generate_lowercase_timezones.go` itself (outside `genrule`)

irfansharif replied:
Perhaps we could try appending to PATH to include `($location @com_github_cockroachdb_crlfmt//:crlfmt)`?

rickystewart replied:
It is really necessary to use `go run` here? That seems like we're not leveraging Bazel for builds correctly. What I would expect is that we compile a new `go_binary` using `generate_lowercase_timezones.go`, and then have a `genrule` that just runs it. That would erase the need to depend on `@go_sdk//:bin/go`. Is there a reason we can't do this?

alan-mas replied:
I was using `go run` as it was the first thing that came to my mind (while checking the actual go:generate)

That is a good suggestion, let me try my luck on that

irfansharif commented on WORKSPACE:
```diff
@@ -85,6 +85,14 @@ go_repository(
     version = "v0.0.0-20200218151345-dad8c97a84f5",
 )
 
+go_repository(
```
Seeing as how this is already specified in go.mod, it's also already present in DEPS.bzl.
```
    go_repository(
        name = "com_github_mmatczuk_go_generics",
        build_file_proto_mode = "disable_global",
        importpath = "github.com/mmatczuk/go_generics",
        sum = "h1:QjLgi/L+MjxysinrA8KkNZLf3cAhTluBoSXUvFeN144=",
        version = "v0.0.0-20181212143635-0aaa050f9bab",
    )
```

rickystewart commented on pkg/roachpb/BUILD.bazel:
```diff
@@ -3,13 +3,13 @@ load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")
 load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 load("//build:STRINGER.bzl", "stringer")
 
+# keep
```
Ah, I don't like this. This will result in Gazelle never updating this target. Why did you make this change?

rickystewart commented on pkg/roachpb/BUILD.bazel:
```diff
@@ -53,6 +54,53 @@ go_library(
+    ],
+)
+
+# keep
```
Same with above. The other instances of `# keep` in this file should be removed unless we have a really, really good reason.

rickystewart commented on pkg/roachpb/genbatch/main.go:
```diff
@@ -11,11 +11,10 @@
 // This file generates batch_generated.go. It can be run via:
```
Since you deleted the `+build gen-batch` tag, this comment is no longer accurate. You should replace it with accurate documentation.

rickystewart commented on pkg/util/timeutil/BUILD.bazel:
```diff
@@ -49,3 +50,21 @@ go_test(
+    tools = [
+        "@com_github_cockroachdb_crlfmt//:crlfmt",
+        "@go_sdk//:ROOT",
+        "@go_sdk//:bin/go",
```
As I said before, I really don't want to link the entire `go` SDK into these `genrules`. Why did we regress on this again? What problem are you trying to solve by doing this?

alan-mas replied:
As code `pkg/util/timeutil/lowercase/main.go` is using `runtime.GOROOT()` we need to tell Bazel which is the path for GOROOT, that is why i found that solution, even that we are already using a `go_binary` `//pkg/util/timeutil/lowercase`

Not sure if you have a better idea on how to handle this.

The other ones works pretty good only with the binary but the main problem in here is that this file is using a file and a run command from GOROOT

rickystewart replied:
This still needs to be fixed.

rickystewart replied:
Again, this still needs to be fixed.

rickystewart commented on pkg/util/timeutil/lowercase/main.go:
```diff
@@ -47,11 +47,11 @@ var lowercaseTimezones = map[string]string{
-	const fileName = "lowercase_timezones.go"
-
+	fileName := *filenameFlag
 	zipdataFile, err := os.Open(filepath.Join(runtime.GOROOT(), "lib", "time", "zoneinfo.zip"))
```
This seems wrong. We shouldn't ever be inferring the location of anything in the Bazel sandbox like this, we should compute it directly and then pass it in as a flag.

alan-mas replied:
Sorry if I am getting incorrectly, but we are passing the `GOROOT` and the `zoneinfo.zip` right on the `tools` of the `genrule`

rickystewart replied:
DON'T pass `GOROOT` in from Bazel. `GOROOT` doesn't exist, as far as Bazel is concerned.

1. Depend on `"@go_sdk//:lib/time/zoneinfo.zip",` in the `genrule`.
2. In the `genrule`, get `$(location @go_sdk//:lib/time/zoneinfo.zip)`, and pass it down to this as an argument.
3. Use that to find where `zoneinfo.zip` is located.

This is how you should always find files you depend on in Bazel. It is NEVER a good idea to find a random directory and try to derive paths from the top directory.

rickystewart commented:
Looks like the (non-`bazel`) build is also failing. What's going on? Do you need help with something?

rickystewart commented on pkg/cmd/roachprod/vm/aws/BUILD.bazel:
```diff
@@ -23,3 +24,25 @@ go_library(
     ],
 )
+
+# Checks: go_library is only expecting .go files as sources.
```
I don't really know what this comment is supposed to mean. What are you trying to express here?

alan-mas replied:
That I am adding this new `main.tf` file but I cannot create a dependency within the main `go_library` as `go_library` is only expecting `.go` files and not `.tf`

rickystewart replied:
Well, the `genrule` is unused right now, so did you mean to use it somewhere?

alan-mas replied:
Yes it is not used yet, I will include it in the next batch of `go:generate` files (mostly go generics)

rickystewart commented on pkg/roachpb/genbatch/main.go:
```diff
@@ -9,13 +9,12 @@
-//    go run -tags gen-batch gen_batch.go
-
-// +build gen-batch
+//    go run -tags genbatch genbatch/main.go
```
Did you test this, are you sure it works?

alan-mas replied:
Yeap, I tried with `make generate` and also manually in my local machine (running command `go run -tags genbatch genbatch/main.go`) ant it is working

rickystewart commented on pkg/security/securitytest/BUILD.bazel:
```diff
@@ -13,3 +14,16 @@ go_library(
 )
+
+bindata(
+    name = "embedded",
```
This looks unused?

alan-mas replied:
nice catch! I just add `# keep` flag and the `bindata` within the main `go_library`

rickystewart commented on pkg/util/log/channel/BUILD.bazel:
```diff
@@ -2,8 +2,23 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library")
+    srcs = [
+        ":gen-channel-go",  # keep
+    ],
+    importpath = "github.com/cockroachdb/cockroach/pkg/util/log/channel",  # keep
```
Is the `# keep` comment necessary here?

alan-mas replied:
Yeap, when I run gazelle is sending this error: 
```
gazelle: finding module path for import github.com/cockroachdb/cockroach/pkg/util/log/channel: exit status 1: go: finding module for package github.com/cockroachdb/cockroach/pkg/util/log/channel
module github.com/cockroachdb/cockroach@latest found (v20.1.10+incompatible), but does not contain package github.com/cockroachdb/cockroach/pkg/util/log/channel
```
and it is removing the import path.

rickystewart reviewed (commented):
Looking pretty good, I have a few questions.

rickystewart commented on BUILD.bazel:
```diff
@@ -34,6 +35,13 @@ load("@bazel_gazelle//:def.bzl", "gazelle")
 # gazelle:resolve proto go google/api/annotations.proto @org_golang_google_genproto//googleapis/api/annotations:go_default_library
 # gazelle:resolve proto prometheus/client_model/metrics.proto @com_github_prometheus_client_model//:client_proto
 # gazelle:resolve proto go prometheus/client_model/metrics.proto @com_github_prometheus_client_model//go
+# gazelle:resolve proto go roachpb/data.proto @cockroach//pkg/roachpb
```
Please alphabetize this list.

rickystewart commented on pkg/util/timeutil/lowercase/main.go:
```diff
@@ -49,11 +48,21 @@ var lowercaseTimezones = map[string]string{
 func main() {
+	var filenameFlag = flag.String("filename", "lowercase_timezones.go", "filename path")
+	var zoneInfo = flag.String("zoneinfo", "zoneinfo.zip", "zoneinfo.zip path")
+	var zipdataFile *os.File
```
This is a pretty wild default value for the flag. :) 

I would just make the default value `""`, then you can check if the string is empty and set it to `GOROOT/lib/time/zoneinfo.zip` if so.

alan-mas replied:
Done =)

alan-mas commented on pkg/roachpb/BUILD.bazel:
```diff
@@ -132,6 +179,7 @@ proto_library(
     ],
 )
 
+# keep
```
If I remove that `# keep` flag, `gazelle` is adding `":roachpb"` as a dependecy and it is creating `cycle in dependency graph` .
Not sure why is adding that `go_library` as a `deps` as it was not there before (even that the `go_library` was already present in code)

Got a better idea on the behavior in `go_proto_library` I already add in main Bazel Build file:
```
# gazelle:resolve go github.com/cockroachdb/cockroach/pkg/roachpb //pkg/roachpb
# gazelle:resolve proto go roachpb/api.proto @cockroach//pkg/roachpb
# gazelle:resolve proto go roachpb/app_stats.proto @cockroach//pkg/roachpb
# gazelle:resolve proto go roachpb/data.proto @cockroach//pkg/roachpb
# gazelle:resolve proto go roachpb/errors.proto @cockroach//pkg/roachpb
# gazelle:resolve proto go roachpb/internal_raft.proto @cockroach//pkg/roachpb
# gazelle:resolve proto go roachpb/io-formats.proto @cockroach//pkg/roachpb
# gazelle:resolve proto go roachpb/metadata.proto @cockroach//pkg/roachpb
```

rickystewart replied:
Hmm, that's really weird. I wonder why it's doing that. OK, it can probably stay for now.

irfansharif replied:
Pretty sure it's because the top-level resolve directives (needed for the code-gen-bootstrapping-thing).

irfansharif commented:
Mind holding off here on merge? I'm working just a bit on this PR, I think there are a few changes we want to make.

alan-mas commented:
> Mind holding off here on merge? I'm working just a bit on this PR, I think there are a few changes we want to make.

Not at all, take your time =) I am checking other `go:generate` in the midtime

irfansharif commented:
I'll encourage you to do it in a separate PR. Feel free to rebase off of this to make your life a bit easier - I'm hoping to adjust a few things in place here.

alan-mas commented:
> I'll encourage you to do it in a separate PR. Feel free to rebase off of this to make your life a bit easier - I'm hoping to adjust a few things in place here.

Yes actually I am working in a new PR for the other files (james shared [this document](https://wiki.crdb.io/wiki/spaces/CRDB/pages/1411744698/Decomposing+changes+and+updating+PRs+during+reviews) with me and it is pretty helpful, so next PR will be apart from this)

irfansharif commented:
<img width="1362" alt="image" src="https://user-images.githubusercontent.com/10536690/108912444-a4aae580-75f6-11eb-9216-12c4194f0b34.png">

Here are just a few more _stringer files we'd want to build through Bazel instead, that we're not doing yet as of this PR.

irfansharif commented:
<img width="1362" alt="image" src="https://user-images.githubusercontent.com/10536690/108912913-44687380-75f7-11eb-9bfb-9bf8a167f7e7.png">

And the list of _generated.go files. These might be already on your radar, but wanted to put it here for posterity. If we've standardized around the naming scheme (_stringer.go, _generated.go), and I think we have, at the top-level BUILD.bazel file we could glob exclude all these files.

```
# gazelle:exclude **/*_generated.go
# gazelle:exclude **/*_stringer.go
```

But that's not possible until we address the remainder of these files. Just a goal to start inching towards.

alan-mas commented:
> <img alt="image" width="1362" src="https://user-images.githubusercontent.com/10536690/108912444-a4aae580-75f6-11eb-9216-12c4194f0b34.png">
> 
> Here are just a few more _stringer files we'd want to build through Bazel instead, that we're not doing yet as of this PR.


Cool, nice catch! I will work on them (should not be impacted by this PR as far as I know) and in worst case scenario I just rebase and that's it =)

irfansharif commented on pkg/roachpb/BUILD.bazel:
```diff
@@ -53,6 +53,52 @@ go_library(
+        "api.go",
+        "app_stats.go",
+        "batch.go",
+        "batch_generated.go",
```
Hm, this is broken. `batch_generated.go` is actually what we want to generate through `genbatch`. I see below we defined a separate `batch-gen.go` target. That's actually not what we want. What we're effectively doing in this PR is using the checked in generated file to generated the next (`batch-gen.go`), which defeats the whole purpose of defining the bootstrap target.

irfansharif replied:
Actually nevermind, we do have to do it this way -- just reminded myself why partitioning out pkg/roachpb into an independently compilable "bootstrap" target is painful. I guess we'll want to do what we did with langgen and soon introduce an out-of-band mechanism to implant generated code (batch-gen.go) back into the source tree (as batch_generated.go).

irfansharif committed:
*,bazel: generate few more files within the sandbox

Touches #57787, we convert over a few more stringer targets and other
instances of files generated using `//go:generate go run <...>`.
files within Bazel.

Of note in this (beast of a) commit is the generation of
roachpb/batch_generated.go, which much like langgen, makes use of
symbols in the package to generate the file later needed to compile the
package. To that end we introduce a stand-alone 'bootstrap' target.

For other files generated using `go run fname.go`, we convert them over
into standalone binaries.

Release note: None

Co-authored-by: Alanmas <acostas.alan@gmail.com>

irfansharif commented:
bors r+

alan-mas commented:
> <img alt="image" width="1362" src="https://user-images.githubusercontent.com/10536690/108912444-a4aae580-75f6-11eb-9216-12c4194f0b34.png">
> 
> Here are just a few more _stringer files we'd want to build through Bazel instead, that we're not doing yet as of this PR.

Correct me if I am wrong, but `pkg/sql/sem/tree/parse_string.go` it is not an auto-generated file (not sure why then it has `-string` on its name

irfansharif commented:
Yup, seems like it's not. I'd focus on the other files then; we can rename that file to be something different.
